nixos/postgresql: fix several issues

aanderse opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
- stop "punishing" long time NixOS users (`stateVersion 17.09` or below) by setting their `dataDir` to `/var/db/postgresql` even though it is their first time installing `postgresql`
- realize we can't cover every possible scenario a sysadmin has for `dataDir` and only provision it when the default value is left (resolves #80827)
- stop using deprecated `PermissionsStartOnly` (partial fix for #53852)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

aanderse committed:
nixos/postgresql: use a standard default value for dataDir

aanderse committed:
nixos/postgresql: conditionally provision data directory with StateDirectory

aanderse committed:
nixos/postgresql: replace deprecated usage of PermissionsStartOnly

aanderse commented:
@GrahamcOfBorg test postgresql postgresql-wal-receiver

aszlig reviewed (approved):
Thanks a lot.

I wonder whether we should move `ExecStartPost` out of the main unit into a separate unit, though, since using `runuser` to switch to the superuser screams "uuuugly!" to me :-D

aanderse commented:
> Thanks a lot.
> 
> I wonder whether we should move `ExecStartPost` out of the main unit into a separate unit, though, since using `runuser` to switch to the superuser screams "uuuugly!" to me :-D

The scope of this PR is perfect in my opinion. I have 3 PRs to follow this one, 1 of which is doing exactly that :smile: 

@GrahamcOfBorg test postgresql postgresql-wal-receiver

aanderse commented:
I run a `matrix-synapse` server so I have an instance of `postgresql` I tested this against, all good for an existing install.

@Ma27 would you be able/willing to test this as well?

Ma27 commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -57,7 +57,10 @@ in
+          The data directory for PostgreSQL. If left as the default value
+          this directory will automatically be created before the PostgreSQL server starts, otherwise
+          the sysadmin is responsible for ensuring the directory exists with appropriate ownership
+          and permissions.
```
Maybe explain what the default is (e.g. with `defaultText`)? The option doesn't have an option-default, so folks reading the manual will be instructed to use the default, but have to check the source to find out what it actually is.

Ma27 reviewed (approved):
LGTM :+1:

aanderse commented:
@GrahamcOfBorg test postgresql postgresql-wal-receiver

aszlig commented on nixos/tests/postgresql-wal-receiver.nix:
```diff
@@ -28,6 +28,9 @@ let
     meta.maintainers = with maintainers; [ pacien ];
 
     machine = { ... }: {
+      systemd.tmpfiles.rules = [
```
```suggestion
      # Needed because this test uses a non-default 'services.postgresql.dataDir'.
      systemd.tmpfiles.rules = [
```

aanderse replied:
Thanks!

Ma27 commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -615,6 +615,17 @@ services.dokuwiki."mywiki" = {
+   <listitem>
+    <para>
+      The <xref linkend="opt-services.postgresql.dataDir"/> option is now set to <literal>"/var/lib/postgresql/${cfg.package.psqlSchema}"</literal> regardless of your
+      <literal>stateVersion</literal>. Users with an existing postgresql install that have a <literal>stateVersion</literal> of <literal>17.09</literal> or below
```
nitpick, but what about
```
<xref linkend="opt-system.stateVersion"/>
```

aanderse replied:
Good catch, thanks!

Ma27 reviewed (approved):
LGTM. Any additional comments? Otherwise I'd consider this mergable :)

edolstra commented:
> stop "punishing" long time NixOS users (stateVersion 17.09 or below) by setting their dataDir to /var/db/postgresql even though it is their first time installing postgresql

But it punishes users who *were* using PostgreSQL, since their PostgresSQL breaks after upgrading to 20.09. (This happened on hydra.nixos.org.) The whole point of `stateVersion` is to prevent that kind of breakage from happening...

aanderse commented:
@edolstra I *hope* I have sufficiently addressed your concerns [here](https://github.com/NixOS/nixpkgs/pull/95294#discussion_r506973706). Please let me know, one way or the other.
